src/Microsoft.Diagnostics.Monitoring/ReversedServerConnectionsSource.cs

```diff
@@ -0,0 +1,194 @@
+                    .Select(c => new DiagnosticsConnection(c))
+                    .ToList()
+                    .AsReadOnly();
+            }
+            finally
+            {
+                _connectionsSemaphore.Release();
+            }
+        }
+
+        /// <summary>
+        /// Accepts connections from the reversed diagnostics server.
+        /// </summary>
+        /// <param name="token">The token to monitor for cancellation requests.</param>
+        private async Task AcceptAsync(CancellationToken token)
+        {
+            // Continuously accept connections from the reversed diagnostics server so
+            // that <see cref="ReversedDiagnosticsServer.AcceptAsync(CancellationToken)"/>
+            // is always awaited in order to to handle new runtime instance connections
+            // as well as existing runtime instance reconnections.
+            while (!token.IsCancellationRequested)
+            {
+                try
+                {
+                    ReversedDiagnosticsConnection connection = await _server.AcceptAsync(token);
+
+                    await _connectionsSemaphore.WaitAsync(token);
+
+                    _connections.Add(connection);
+
+                    _connectionsSemaphore.Release();
+
+                    if (null != _newConnectionSource)
+                    {
+                        _newConnectionSource.TrySetResult(null);
+                    }
+                }
+                catch (OperationCanceledException)
+                {
+                }
+            }
+        }
+
+        /// <summary>
+        /// Tests a diagnostics connection and removes it if the associated runtime instance is no longer active.
+        /// </summary>
+        private async Task PruneConnectionAsync(ReversedDiagnosticsConnection connection, CancellationToken token)
+        {
+            // Cancel event pipe processing within reasonable amount of time to remove potentially unresponsive processes.
+            using CancellationTokenSource timeoutSource = new CancellationTokenSource(TimeSpan.FromSeconds(5));
+
+            // Try to get ProcessInfo information to test the connection
+            try
+            {
+                DiagnosticsClient client = new DiagnosticsClient(connection.Endpoint);
+
+                // NOTE: This test does not need to necessarily use the event pipe providers and collection.
+                // It only needs to send some type of request and receive a response from the runtime instance
+                // to verify that the diagnostics pipe/socket is viable and the runtime instance is still connected.
+                await using DiagnosticsEventPipeProcessor process = new DiagnosticsEventPipeProcessor(PipeMode.ProcessInfo);
+
+                using CancellationTokenSource linkedSource = CancellationTokenSource.CreateLinkedTokenSource(token, timeoutSource.Token);
+
+                await process.Process(client, 0, Timeout.InfiniteTimeSpan, linkedSource.Token);
```

Comment:
Do we know how much overhead this adds?

Reply:
While we might be able to do that for sockets, named pipes don't have that ability as far as I know. It's my understanding that you have to try to send or read something to/from the pipe to check if it's valid/connected (e.g. ReadFile + GetLastError). I'll see if ReadFile/WriteFile support reading/writing 0 bytes and only return an error if the pipe is broken/bad state/etc. If that works, then we can push this down lower.

Reply:
For what it's worth, I use a 0 byte non-blocking read in the runtime to test the "readiness" of namedpipes.  I would imagine something similar should work here, but with a blocking read.

Reply:
> For what it's worth, I use a 0 byte non-blocking read in the runtime to test the "readiness" of namedpipes. I would imagine something similar should work here, but with a blocking read.

Would you mind pointing me to that?

Reply:
https://github.com/dotnet/runtime/blob/02d182f78d613a2a5689e7b08b6997d4cf6f1265/src/coreclr/src/debug/debug-pal/win/diagnosticsipc.cpp#L274-L306

It's directly using the Win32 APIs, so I don't know how it will map to the C# APIs.

Reply:
This is now using IIpcEndpoint.CheckTransport since that now verifies that the existing stream is still connected with the runtime instance. It uses PeekNamedPipe for Windows named pipes and non-blocking zero-byte socket writing for non-Windows.